examples/small-coalton-programs/src/brainfold.lisp

```diff
@@ -0,0 +1,341 @@
+;;;;
+;;;;
+;;;; A Brainfold/Brainf*** compiler implemented in Coalton
+;;;;
+;;;; The compiler handles all standard bf commands:
+;;;; ( > < + - . , [ ] ).
+;;;;
+;;;;
+;;;; Run Brainfold programs with (run-program "+++++[-]+++")
+;;;;
+;;;; or
+;;;;
+;;;; (run-file "/path/to/your/file.bf")
+;;;;
+;;;; try (coalton (hello-world)) in the REPL!
+;;;;
+
+
+(cl:defpackage #:brainfold
+  (:use
+   #:coalton
+   #:coalton-prelude)
+  (:local-nicknames
+   (#:vec #:coalton-library/vector)
+   (#:iter #:coalton-library/iterator)
+   (#:cell #:coalton-library/cell)
+   (#:char #:coalton-library/char)
+   (#:str #:coalton-library/string)
+   (#:list #:coalton-library/list)
+   (#:arith #:coalton-library/math))
+
+  (:export
+   #:eval
+   #:run-program
+   #:run-file
+   #:hello-world))
+
+;;;
+;;; Environment (Vector, Pointer, Print Buffer)
+;;;
+
+(in-package #:brainfold)
+
+(named-readtables:in-readtable coalton:coalton)
+
+(coalton-toplevel
+
+  ;;
+  ;; Initializing a vector with 0's, ideally 30k according to the
+  ;; specifications
+  
+  (declare bf-array-size (UFix))
+  (define bf-array-size 1000)
+
+  (define (generate-bf-array)
+    "Initializes the brainfold array."
+    (let v = (vec:new))
+    (vec:extend! v (iter:repeat-for 0 bf-array-size))
+    v)
+ 
+  (define bf-array (generate-bf-array))
+
+  (define (reset-bf-array)
+    "Resets the brainfold array."
+    (vec:clear! bf-array)
+    (vec:extend! bf-array (iter:repeat-for 0 bf-array-size))
+    bf-array)
+
+  ;;
+  ;; The pointer to the current Brainfold cell
+  ;;
+  
+  (declare pointer (Cell UFix))
+  (define pointer (cell:new 0))
+
+  ;;
+  ;; Accessing Brainfold cells using pointer
+  ;;
+  
+  (define (nth-bf-cell index)
+    "Finds the vector entry associated with the given index."
+    (vec:index index bf-array))
+
+  (define (current-bf-cell-value)
+    "Returns the value of the bf-cell indexed by the pointer."
+    (unwrap (nth-bf-cell (cell:read pointer))))
+
+  ;;
+  ;; Print Buffer for printing
+  ;;
+
+  (define print-buffer (cell:new ""))
+  
+  ;;
+  ;; Clean Environment
+  ;;
+  
+  (define (reset-bf)
```

Comment:
Instead of making the BF state global, you might instead want to try

```
(define-type State
  (State (Cell String) (Cell UFix) (Vector UFix)))

(define (state-print-buffer (State b _ _))
  b)

(define (state-pointer (State _ p _))
  p)

(define (state-memory (State _ _ m))
  m)
```

That way all of these functions can take and return the state as arguments:

```
(declare reset-bf (State -> State))
(define (reset-bf s)
  (cell-write! (state-print-buffer s) "")
  (cell-write! (state-pointer s) 0)
  (reset-bf-array (state-memory s))
  s)
```

The big downside is that it takes away from the pedagogy of the example. Now your code gets riddled with accessors everywhere. Worth it for a minimal example? Not sure.